src/data/articles/scala-3-the-evolution-of-a-macro.mdx

```diff
@@ -0,0 +1,1829 @@
+tags: [fp, scala, scala-3, macros]
+title: "Scala 3: The Evolution of a Macro"
+---
+## Introduction
+
+In this post we are going to explore how Scala 3's macros can be used to simplify the use of a semi-realistic and boilerplaty tracing library.
+We'll see how we can reduce boilerplate using different levels of macro implementations, starting from simple but minimally helpful, and evolving our macros into something more complex but almost magical.
+
+Each step in this evolution will use various macro capabilities and we'll see how to leverage them to reduce boilerplate in our code.
+
+## Disclaimer
+
+This post is long, going through a lot of code. Please take your time reading it, and let each part sink in before moving on.
+
+Throughout the code examples basic understanding of Scala 3's macro APIs is assumed. Things like quotes/splices, `Expr`, `Tree`, `Symbol`, etc., should be familiar to the reader. You can think of this blog as a supplement to Rock the JVM's [metaprogramming course](/courses/scala-macros-and-metaprogramming). For an in-depth introduction to macros you can read [this](scala-3-macros-comprehensive-guide) article.
+
+## Setup
+
+All the code from the post is available in <a href={frontmatter.repositoryUrl}>this</a> repo. The project is set up using sbt. If you want to follow along in a REPL, please use the project's sbt console:
+```
+sbt console
+```
+
+This will set up the imports required for code snippets to work.
+
+You can use:
+```
+sbt run
+```
+To choose any of the runnable examples in the code.
+
+The project uses a couple compilation flags:
+- `-Xprint:postInlining` - prints out the code that the compiler generated after expanding macros, this is useful for macro inspection/debugging, and can be turned off if the output is too noisy (disabled in the console)
+- `-experimental` - enables all of Scala's experimental features without requiring the `@experimental` annotation
+
+The code samples below assume the presence of some helper classes, mostly related to the fictional tracing library.
+The full code can be found in the <a href={frontmatter.repositoryUrl}>repo</a>, we won't be showing it in detail here.
+
+Due to how macros work, defining and using macros must happen in separate files. If you're running code in the REPL you'll need to write the code for the macro, evaluate it, and only then try to use it.
+
+With the setup out of the way, we are ready to dive in. But first, let's take a look at a baseline implementation without macros, so that we can see what kind of code we want to automate.
+
+##  Baseline
+
+Our goal is to trace a block of code with some metadata that is provided by the user. That is, we want to surround a call to the block of code with informative calls to the tracing library. Here's our baseline implementation:
+
+```scala wrap=false
+import scala.util.control.NonFatal
+
+object Baseline:
+  def trace0[A](level: TraceLevel, // 1
```

Comment:
replace indentation syntax and use curly braces - copying and pasting is a mess otherwise

Reply:
I know you're diligent about it, but other people's settings may screw this up. I know because I've seen it happen, and replaced all indentation syntax across the entire site. 

Thanks for understanding, and sorry about the hassle

Reply:
:heavy_check_mark: